squidpy/im/_container.py

```diff
@@ -619,25 +698,58 @@ def generate_spot_crops(
         Yields
         ------
         If ``return_obs = True``, yields a :class:`tuple` ``(crop, obs_name)``. Otherwise, yields just the crops.
-        The type of the crops depends on ``as_array``.
+        The type of the crops depends on ``as_array`` and the number of dimensions on ``squeeze``.
         """
         self._assert_not_empty()
         _assert_positive(spot_scale, name="scale")
         _assert_spatial_basis(adata, spatial_key)
-        library_id = Key.uns.library_id(adata, spatial_key=spatial_key, library_id=library_id)
 
+        # limit to obs_names
         if obs_names is None:
             obs_names = adata.obs_names
         obs_names = _assert_non_empty_sequence(obs_names, name="observations")
-
         adata = adata[obs_names, :]
-        spatial = adata.obsm[spatial_key][:, :2]
 
         scale = self.data.attrs.get(Key.img.scale, 1)
-        diameter = Key.uns.spot_diameter(adata, spatial_key=spatial_key, library_id=library_id) * scale
-        radius = int(round(diameter // 2 * spot_scale))
+        spatial = adata.obsm[spatial_key][:, :2]
+
+        if library_id is None:
+            try:
+                library_id = Key.uns.library_id(adata, spatial_key=spatial_key, library_id=None)
+                obs_library_ids = [library_id] * adata.n_obs
+            except ValueError as e:
+                if "Unable to determine which library id to use" in str(e):
+                    raise ValueError(
+                        str(e)
+                        + " Or specify a key in `adata.obs` containing a mapping from observations to library ids."
```

Comment:
I tried to make the error message a bit clearer here, as the preferred way of specifying library_id should be an entry in adata.obs (for z-stacks anyways). Not sure if this is the best way of concatenating error messages...

Reply:
Thanks, looks great!